aten/src/TH/generic/THTensorApply.hpp

```diff
@@ -1,41 +1,323 @@
+    /* The following strategy is not easy to understand.
+     * 1. Collapse the dimension of the tensors in order to decrease the number of nested loops.
+     * 2. Calculate the numbers of elements allocated in each thread and the line index of the first one.
+     * 3. Calculate the memory offset of the first element and the indexes in each dimension of the
+     *    first one.
+     * 4. iterate all elements in each thread. update the indexes in each dimension of the rest.
+    */ \
+    int TH_TENSOR_APPLY_hasFinished = 0; \
+    int64_t TH_TENSOR_dim_index = 0; \
+    /*step 1*/ \
+    __TH_TENSOR_APPLYX_PREAMBLE(TYPE2, TENSOR2, -1, 1) \
+    __TH_TENSOR_APPLYX_PREAMBLE(TYPE1, TENSOR1, -1, 1) \
+    if (0 == TH_TENSOR_APPLY_hasFinished) { \
+      auto TENSOR1##_i_local = TENSOR1##_i; \
+      auto TENSOR2##_i_local = TENSOR2##_i; \
+      auto TENSOR1##_data_local = TENSOR1##_data; \
+      auto TENSOR2##_data_local = TENSOR2##_data; \
+      at::parallel_for(0, SIZE, THRESHOLD, [&](int64_t begin, int64_t end) { \
+        auto TENSOR1##_i = TENSOR1##_i_local; \
+        auto TENSOR2##_i = TENSOR2##_i_local; \
+        auto TENSOR1##_data = TENSOR1##_data_local; \
+        auto TENSOR2##_data = TENSOR2##_data_local; \
+        /*step 2*/ \
+        ptrdiff_t line_index_start = begin; \
+        ptrdiff_t line_seg_length = (end - begin); \
+        /* step 3*/ \
+        __TH_TENSOR_APPLYX_CAL_MEMORY_OFFSET(TENSOR2); \
+        __TH_TENSOR_APPLYX_CAL_MEMORY_OFFSET(TENSOR1); \
+        TENSOR2##_data += TENSOR2##_memory_offset; \
+        TENSOR1##_data += TENSOR1##_memory_offset; \
+        ptrdiff_t count = 0; \
+        ptrdiff_t TENSOR2##_start =  TENSOR2##_counter_tmp[TENSOR2##_dim-1]; \
+        ptrdiff_t TENSOR1##_start =  TENSOR1##_counter_tmp[TENSOR1##_dim-1]; \
+        /* step 4*/ \
+        while (count < line_seg_length) { \
+          for (TENSOR2##_i=TENSOR2##_start, TENSOR1##_i = TENSOR1##_start; ((count < line_seg_length) && (TENSOR2##_i < TENSOR2##_size) && (TENSOR1##_i < TENSOR1##_size)); ++TENSOR2##_i, ++TENSOR1##_i, ++count) { \
+            CODE \
+            TENSOR2##_data += TENSOR2##_stride; \
+            TENSOR1##_data += TENSOR1##_stride; \
+          } \
+          if (count < line_seg_length) { \
+            __TH_TENSOR_APPLYX_UPDATE_COUNTERS_PARALLEL(TENSOR2); \
+            __TH_TENSOR_APPLYX_UPDATE_COUNTERS_PARALLEL(TENSOR1); \
+          } \
+        } \
+        if (TENSOR1##_counter_tmp != NULL) { \
+          THFree(TENSOR1##_counter_tmp); \
+        } \
+        if (TENSOR2##_counter_tmp != NULL) { \
+          THFree(TENSOR2##_counter_tmp); \
+        } \
+      }); \
+    } \
+    if (TENSOR2##_counter != NULL) { \
+      THFree(TENSOR2##_counter); \
+    } \
+    if (TENSOR1##_counter != NULL) { \
+      THFree(TENSOR1##_counter); \
+    } \
+  } \
+}
+
+#define TH_TENSOR_APPLY3_PARALLEL(SIZE, CONTIG1, CONTIG2, CONTIG3, TYPE1, TENSOR1, TYPE2, TENSOR2, TYPE3, TENSOR3, CODE, THRESHOLD) \
```

Comment:
I assume this code is a copy-paste from the original macros with openmp replaced with at::parallel_for

Reply:
correct, with some minimum modifications caused by introduction of lambdas